src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageSearch/PackageSearchRunner.cs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Protocol.Core.Types;
+using NuGet.Protocol;
+using NuGet.Commands;
+
+namespace NuGet.CommandLine.XPlat
+{
+    internal static class PackageSearchRunner
+    {
+        /// <summary>
+        /// Runs the search operation asynchronously using the provided parameters.
+        /// </summary>
+        /// <param name="sourceProvider">The provider for package sources.</param>
+        /// <param name="packageSearchArgs">Package search arguments</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        public static async Task<int> RunAsync(
+            IPackageSourceProvider sourceProvider,
+            PackageSearchArgs packageSearchArgs,
+            CancellationToken cancellationToken)
+        {
+            IList<PackageSource> listEndpoints;
+
+            try
+            {
+                listEndpoints = GetPackageSources(packageSearchArgs.Sources, sourceProvider);
+            }
+            catch (ArgumentException ex)
+            {
+                packageSearchArgs.Logger.LogError(ex.Message);
+                return 1;
+            }
+
+            WarnForHTTPSources(listEndpoints, packageSearchArgs.Logger);
+
+            Func<PackageSource, Task<IEnumerable<IPackageSearchMetadata>>> searchPackageSourceAsync =
+                packageSearchArgs.ExactMatch
+                ? packageSource => GetPackageAsync(packageSource, packageSearchArgs, cancellationToken)
+                : packageSource => SearchAsync(packageSource, packageSearchArgs, cancellationToken);
+
+            Dictionary<Task<IEnumerable<IPackageSearchMetadata>>, PackageSource> searchRequests = new();
+
+            foreach (var packageSource in listEndpoints)
+            {
+                Task<IEnumerable<IPackageSearchMetadata>> searchTask = searchPackageSourceAsync(packageSource);
+                searchRequests.Add(searchTask, packageSource);
+            }
+
+            IPackageSearchResultRenderer packageSearchResultRenderer;
+            packageSearchResultRenderer = new PackageSearchResultRendererTable(packageSearchArgs);
+            packageSearchResultRenderer.Start();
+
+            while (searchRequests.Count > 0)
+            {
+                Task<IEnumerable<IPackageSearchMetadata>> completedTask = await Task.WhenAny(searchRequests.Keys);
```

Comment:
What's the output if any of these tasks throw because say the server response took too long, or credentials could not be acquired?

Would we handle the failure scenario gracefully? Should our exit code be different if a failure is encountered? What if one of the sources fails? What if all of them do?

Reply:
I added catch statements to the renderer so that if any of the tasks throw an exception, that could be thrown by the NuGet.Protocol SearchAsync method is handled.

Reply:
The more I read it, the more confident I get that the renderer should not be handling errors. 

When you add another renderer, you need to handle that scenario twice.

Reply:
Fixed. I moved the running the search tasks to the search runner. The renderer takes Ienumerable<IpackagesearchMetadata>